### PR TITLE
Feature: Extended Time data

### PR DESCRIPTION
**Pull request by andrebrait**
Add parsing of extended time data 
    - Detects and parses the EXT_TIME part of the file header
    - `java.util.Date` fields added or improved:
        - `atime`, `ctime` and `arctime` are now parsed, precision is milliseconds
        - `mtime` now has a maximum precision of milliseconds instead of seconds, when EXT_TIME is present
    - Corresponding `FileTime` fields added
        - `lastModifiedTime`, `lastAccessTime` and `creationTime` fields added 
        - precision for those is 100ns intervals

### CODE REVIEW

**Comment by andrebrait** [comment]
Instead of having both Date and FileTime fields, I think I'm gonna keep only FileTime and make the getters and setters for the Date fields just interact get/set to the FileTime fields.

WDYT?

**Comment by gotson** [comment]
Hi, could you add a description of what the PR is about? It would be good to understand what this is _without_ having to read the code. Thanks

**Comment by andrebrait** [comment]
> Hi, could you add a description of what the PR is about? It would be good to understand what this is _without_ having to read the code. Thanks

My bad. I thought I had written it clearly enough, but re-reading the description, I see it's a bit confusing. I'll be editing it.

**Comment by andrebrait** [comment]
> Hi, could you add a description of what the PR is about? It would be good to understand what this is _without_ having to read the code. Thanks

Done. Please check it out.

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (andrebrait) [inline src/main/java/com/github/junrar/rarfile/FileHeader.java:270]: @gotson This is here just we I can reuse the block below without changing anything. I figured out how to make it into a function and I'm gonna refactor it so it'll be a lot cleaner.

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (andrebrait) [inline src/main/java/com/github/junrar/rarfile/FileHeader.java:78]: @gotson given the `FileTime` fields above are available and contain the most accurate precision which reflects the extended time format better (and it also makes this more like `ZipEntry`), I'm thinking it we can get rid of the `java.util.Date` fields and just get them from the `FileTime` equivalents when calling the getters (documenting the change accordingly, of course).

Ex.:

```java
public Date getMTime() {
    return toDate(lastModifiedTime);
}
```

**Comment by gotson** [comment]
I've never heard of the EXT_TIME header. Would you be able to share some context about what this is, how it can be used (outside this project), and what your proposed change can help do that couldn't be done before?

**Comment by andrebrait** [comment]
@gotson I think I mistakenly called it EX_TIME _header_ but what I actually meant was EXT_TIME _section_ of the header. 

I think WinRAR calls it "High precision time format".

Time in RAR 4 and below is stored as an MS-DOS date/time 32-bit integer, which JUnrar decodes into a (local) `java.util.Date` (the `mTime` field).
This format can only store up to the seconds of a given time, and it can only store _even_ values for seconds, not _odd_ values (so `2022-03-04T10:00:55` is stored as `2022-03-04T10:00:54`, for example).

The EXT_TIME flag in the header signals that the header contains additional data about file times.

If the File Header contains this flag, it has a section which contains:
- additional data* for the `mTime` field
- creation time (in MS-DOS format) and additional data* for it
- last access time (in MS-DOS format) and additional data* for it

Each _additional data_ section comprises:
- the fractions of a second for that time, in 100ns increments
- a flag for correcting the MS-DOS date/time format if the second was actually _odd_

Parsing this section of the file header enables it to parse dates correctly using all information available in the RAR file.
So, basically, it enables us to increase the precision of the data it already parsed (mTime) from seconds to seconds + 100ns increments, and it enables us to also read the creation and access times for the entry.

For example, the following entry (output from the UnRAR CLI utility):

```
        Name: files\test\lorem-ipsum.txt
    Modified: 2022-02-23 10:25:55,206205200
     Created: 2022-02-23 10:19:22,854708200
    Accessed: 2022-03-02 18:45:18,695090800
 ```
 
 before this patch, is parsed as (notice the seconds which went from 55 to 54 because of the MS-DOS time/date format):
 
 ```
        Name: files\test\lorem-ipsum.txt
    Modified: 2022-02-23 10:25:54
     Created: 
    Accessed: 
 ```

And after this patch, it's parsed as:

```
        Name: files\test\lorem-ipsum.txt
    Modified: 2022-02-23 10:25:55,206205
     Created: 2022-02-23 10:19:22,854708
    Accessed: 2022-03-02 18:45:18,695090
 ```
 
 (the last 3 digits are missing because I've only coded the precision to be up to microseconds, which I'm fixing today and it'll have access to the full 100ns increments)

**Comment by andrebrait** [comment]
There was a `// TODO rartime -> extended` section there, so I assumed you knew about this, which is why I didn't go for a more detailed explanation right away. My bad.

**Comment by andrebrait** [comment]
And this is where I took most of the things here, including the pseudo-code:

https://codedread.github.io/bitjs/docs/unrar.html

See the [ExtTime structure section](https://codedread.github.io/bitjs/docs/unrar.html#ext-time-structure).

**Comment by andrebrait** [comment]
@gotson I just pushed some improvements, including now it being fully compatible with 100ns intervals.

I think I ended up pushing some small optimization around array copies 😅 
It's just stuff addressing the difference between C++ and Java strings and not needing to copy arrays just to make strings in Java, but it was intended for another branch.

**Comment by andrebrait** [comment]
I reverted most things, leaving only the functionality itself here. That way it's easier to review.

**Comment by gotson** [comment]
> There was a `// TODO rartime -> extended` comment in there, so I assumed you knew about this, which is why I didn't go for a more detailed explanation right away. My bad.

I am just the maintainer of the project, the code is more than 15 years old :)




> I reverted most things, leaving only the functionality itself here. That way it's easier to review.

Thanks, i was going to ask for splitting up the various changes in multiple PRs, so it's easier to review and merge while keeping a nice commit history.

---

Thanks a lot for the PR, i will review it when i have more time.

File: src/test/java/com/github/junrar/ArchiveTest.java
Comment (gotson) [inline src/test/java/com/github/junrar/ArchiveTest.java:102]: Wouldn't this change the behaviour for subsequent tests also ?

**Comment by gotson** [comment]
> Instead of having both Date and FileTime fields, I think I'm gonna keep only FileTime and make the getters and setters for the Date fields just interact get/set to the FileTime fields.
> 
> WDYT?

I think `FileTime` would make more sense.

That would be a breaking change though, make sure you add `BREAKING CHANGE:` in the commit description, it would be picked up by semantic release and bump the major version accordingly ([example](https://github.com/junrar/junrar/commit/e6f461b60875e582ac54ee4b8b3a23744d5d97c0)).

File: src/test/java/com/github/junrar/ArchiveTest.java
Comment (andrebrait) [reply src/test/java/com/github/junrar/ArchiveTest.java:102]: That's what line 126 is for.
As long as we're not running the tests in parallel, it should be fine.

I can also add the TimeZone to the constructor of Archive or try to pass it some other way if you think that's better.

Comment (gotson) [reply src/test/java/com/github/junrar/ArchiveTest.java:102]: Maybe we could use https://junit-pioneer.org/docs/default-locale-timezone/#defaulttimezone

**Comment by andrebrait** [comment]
> I think `FileTime` would make more sense.
> 
> That would be a breaking change though, make sure you add `BREAKING CHANGE:` in the commit description, it would be picked up by semantic release and bump the major version accordingly ([example](https://github.com/junrar/junrar/commit/e6f461b60875e582ac54ee4b8b3a23744d5d97c0)).

I actually already did that. There's no breakage because the getters and setters are kept, but they get/set the FileTime fields in the appropriate ways.

**Comment by gotson** [comment]
> > I think `FileTime` would make more sense.
> > That would be a breaking change though, make sure you add `BREAKING CHANGE:` in the commit description, it would be picked up by semantic release and bump the major version accordingly ([example](https://github.com/junrar/junrar/commit/e6f461b60875e582ac54ee4b8b3a23744d5d97c0)).
> 
> I actually already did that. There's no breakage because the getters and setters are kept, but they get/set the FileTime fields in the appropriate ways.

But you don't expose it externally though. Should it be?

File: src/test/java/com/github/junrar/ArchiveTest.java
Comment (andrebrait) [reply src/test/java/com/github/junrar/ArchiveTest.java:102]: Sure. I can try that.

**Comment by andrebrait** [comment]
> But you don't expose it externally though. Should it be? 

What do you mean? There are getter/setter methods for the FileTime fields that only get/set the objects themselves without any conversion.

FileTimes are immutable, so that's safe to do.

**Comment by gotson** [comment]
> > But you don't expose it externally though. Should it be?
> 
> What do you mean? There are getter/setter methods for the FileTime fields that only get/set the objects themselves without any conversion.
> 
> FileTimes are immutable, so that's safe to do.

Ah! I didn't see you added new getter/setter and also kept the original.

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (gotson) [inline src/main/java/com/github/junrar/rarfile/FileHeader.java:268]: Could probably use double append on those, since the rest will be done in your other PR.

**Comment by gotson** [comment]
> > > But you don't expose it externally though. Should it be?
> > 
> > 
> > What do you mean? There are getter/setter methods for the FileTime fields that only get/set the objects themselves without any conversion.
> > FileTimes are immutable, so that's safe to do.
> 
> Ah! I didn't see you added new getter/setter and also kept the original.

I'm wondering if the old getter/setter should be marked as deprecated. Thoughts?

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (gotson) [inline src/main/java/com/github/junrar/rarfile/FileHeader.java:252]: Are you sure about this one? It should be the second bit, and this is seems to check the 4th instead, no? I'm not very fluent in bitwise operation, but this also differs from the pseudo code from the website you referenced, where they use `4` and not `0x4`.

File: src/test/java/com/github/junrar/ArchiveTest.java
Comment (gotson) [reply src/test/java/com/github/junrar/ArchiveTest.java:102]: Line 126 would not always be triggered if you have an exception, no?

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (andrebrait) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java:252]: 4 and 0x4 are the same thing.

4 * 10^0 = 4 * 16^0 = 1*2^2 + 0*2^1 + 0* 2^0

So 4 here is 0100, which is the second bit.

Comment (gotson) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java:252]: BTW do you think you could add a test for an archive that needs the 1s correction?

File: src/test/java/com/github/junrar/ArchiveTest.java
Comment (andrebrait) [reply src/test/java/com/github/junrar/ArchiveTest.java:102]: Yes, but if you have an exception the test fails anyway and the build stops, unless we're running the tests in the suite in parallel.

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (andrebrait) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java:252]: That branch is covered in the archives I used for testing, isn't it?

The archive I added for testing needs this correction. Without the correction its modification time ends in 54 seconds. With it, it ends in 55, and that's checked in the tests.

Or do you mean a specific test for that?

Comment (gotson) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java:252]: my bad, i didn't check the coverage 🤦🏻

File: src/test/java/com/github/junrar/ArchiveTest.java
Comment (gotson) [reply src/test/java/com/github/junrar/ArchiveTest.java:102]: better plan for the future if we can!

Comment (andrebrait) [reply src/test/java/com/github/junrar/ArchiveTest.java:102]: I have to check how that solution works because it might not be different than a try-finally block, which won't solve parallel test execution

Comment (andrebrait) [reply src/test/java/com/github/junrar/ArchiveTest.java]: @gotson done :)

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (andrebrait) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java]: Done

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (theobisproject) [inline src/main/java/com/github/junrar/rarfile/FileHeader.java:239]: I personally would prefer to use the Java TimeUnit class to do this conversion here and in the else below. Code would look like this then
```suggestion
                seconds = TimeUnit.MILLISECONDS.toSeconds(baseTime.toMillis());
```
I leave it up to you since this is only cosmetic

Comment (andrebrait) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java:239]: I think FileTime has a `to(TimeUnit)` method that I could just have used for that. It slipped my mind.

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (gotson) [inline src/main/java/com/github/junrar/rarfile/FileHeader.java:242]: How about something like:
```java
seconds = getDateDos(Raw.readIntLittleEndian(fileHeader, position)) / TimeUnit.SECONDS.toMillis(1L);
```

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (gotson) [inline src/main/java/com/github/junrar/rarfile/FileHeader.java:254]: And how about:
```java
nanos += TimeUnit.SECONDS.toNanos(1L);
```

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (andrebrait) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java:242]: I guess this would be more idiomatic.

```java
seconds = TimeUnit.MILLISECONDS.toSeconds(getDateDos(Raw.readIntLittleEndian(fileHeader))```

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (andrebrait) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java:254]: Sure. Gonna change it later today

File: src/main/java/com/github/junrar/rarfile/FileHeader.java
Comment (gotson) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java:242]: nice

Comment (andrebrait) [reply src/main/java/com/github/junrar/rarfile/FileHeader.java]: done

**Review by gotson — APPROVED** [review]
LGTM

**Comment by gotson** [comment]
@theobisproject i would be happy to have your updated review before merging :)

**Comment by theobisproject** [comment]
LGTM

**Pull request merged**
